### PR TITLE
luci-app-usteer: First version of luci-app-usteer

### PR DESCRIPTION
**Pull request by Ramon00**
This implements the 1st version of luci-app-usteer. 3 pages, 1 for settings 1 for status of connected clients and 1 for the hearing map.

Note this is a resubmit (see https://github.com/openwrt/luci/pull/6658 for review) .

### CODE REVIEW

**Comment by ptpt52** [comment]
nice.

**Comment by McGiverGim** [comment]
The comment of the commit is wrong, it seems it contains comments from all the commits from the first version. Apart of that, some strange indentations but it's ok to me.

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+				E('th', { 'class': 'th' }, _('rrm_nr mac')),
+				E('th', { 'class': 'th' }, _('rrm_nr hex'))
+				
+			])
+		]);
+	
+		var connectioninfo_table_entries =[];
+		for(var wlan in Localinfo) {
+			connectioninfo_table_entries.push([
+							'<nobr>'+wlan+'</nobr>',
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:112]: Is `wlan` the network SSID? If so, please change the code into `'<nobr>%h</nobr>'.format(wlan),` to HTML escape the value in order to avoid XSS.

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+							Localinfo[wlan]['max_assoc'],
+							Localinfo[wlan]['roam_events']['source'],
+							Localinfo[wlan]['roam_events']['target'],
+							Localinfo[wlan]['rrm_nr'][0],
+							Localinfo[wlan]['rrm_nr'][2]
+						]);
+				}	
+		for(var wlan in Remoteinfo) {
+			connectioninfo_table_entries.push([
+							'<nobr>'+wlan+'</nobr>',
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:128]: ditto

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+								}
+								if ((String(hosts[macUp]['name']).length>0) && (typeof hosts[macUp]['name'] !== 'undefined')) {
+									foundname=hosts[macUp]['name'];
+								}
+							}
+							hostl=hostl+ foundname+'&emsp;';
+				}
+			}		
+			compactconnectioninfo_table_entries.push([
+								'<nobr>'+wlan+'</nobr>',
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:180]: ditto

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+								}
+								if ((String(hosts[macUp]['name']).length>0) &&	(typeof hosts[macUp]['name'] !== 'undefined')) {
+									foundname=hosts[macUp]['name'];
+								}
+							}
+							hostl=hostl+ foundname+'&emsp;';
+				}
+			}	
+			compactconnectioninfo_table_entries.push([
+								'<nobr>'+wlan+'</nobr>',
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:208]: ditto

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+							var macUp=String(mac).toUpperCase()
+							if (typeof hosts[macUp] !== 'undefined') {
+								if ((String(hosts[macUp]['ipaddrs'][0]).length>0) && (typeof hosts[macUp]['ipaddrs'][0] !== 'undefined')) {
+									foundname=hosts[macUp]['ipaddrs'][0]; 
+								}
+								if ((String(hosts[macUp]['name']).length>0) &&	(typeof hosts[macUp]['name'] !== 'undefined')) {
+									foundname=hosts[macUp]['name'];
+								}
+							}
+							hostl=hostl+ foundname+'&emsp;';
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:204]: There's a chance that the hostname is bogus/malicious. Please change the expression to `hostl += '%h\u2003'.format(foundname);`

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+		);
+		for(var mac in Clients) {
+			var maciphost='';
+			maciphost='Mac: '+mac;
+			var macUp=String(mac).toUpperCase()
+				if (typeof hosts[macUp] !== 'undefined') {
+					if (typeof hosts[macUp]['ipaddrs'] !== 'undefined') 
+						maciphost=maciphost+'&emsp;IP: '+hosts[macUp]['ipaddrs'];
+					if (typeof hosts[macUp]['name'] !== 'undefined') 
+						maciphost=maciphost+'&emsp;Host: '+hosts[macUp]['name'];
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:233]: There's a chance that the hostname is bogus/malicious. Please change the expression to `maciphost += '\u2003Host: %h'.format(hosts[macUp]['name']);`

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+					E('th', { 'class': 'th' }, _('connected')),
+					E('th', { 'class': 'th' }, _('signal'))
+				])
+			]);
+		
+			var client_table_entries =[];
+			for(var wlanc in Clients[mac]) {
+				client_table_entries.push([
+					'<nobr>'+wlanc+'</nobr>', 
+					(String(Clients[mac][wlanc]['connected']).valueOf()==String("true").valueOf()) ? "True" : "",
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:250]: Should be just
`Clients[mac][wlanc]['connected'] === true ? 'True' : ''`

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+		body.appendChild(
+			E('h3', 'Hearing map')
+		);
+		for(var mac in Clients) {
+			var maciphost='';
+			maciphost='Mac: '+mac;
+			var macUp=String(mac).toUpperCase()
+				if (typeof hosts[macUp] !== 'undefined') {
+					if (typeof hosts[macUp]['ipaddrs'] !== 'undefined') 
+						maciphost=maciphost+'&emsp;IP: '+hosts[macUp]['ipaddrs'];
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:231]: Please replace `&emsp;` with `\u2003`

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/settings.js [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+	o.value('0','0 Fatal');
+	o.value('1','1 info');
+	o.value('2','2 Verbose');
+	o.value('3','3 Some debug');
+	o.value('4','4 network packet info');
+	o.value('5','5 all debug messages');
+	o.rmempty = false;
+	o.editable = true;
+
+	o = s.option(form.Value, 'max_neighbour_reports', _('max neighbour reports'), _('Maximum number of neighbor reports set for a node'));
```
Comment (jow-) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/settings.js:34]: Please capitalize option titles, like it is done elsewhere in LuCI.

**Review by jow- — CHANGES_REQUESTED** [review]
Great work! I added a number of remarks. Please don't open a new PR but keep updating this one.

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 1 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:112]: No, its the IP+interface name so something like "192.168.64.1#hostapd.phy0-ap0". Note this table only contains information about the users own network, so XSS should not be an issue. I can change it to that suggestion, but I do not think it is really needed. Let me know what you prefer

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 5 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:204]: yes let me do this.

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 6 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:233]: agree

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 7 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:250]: ok

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 8 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:231]: ok

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+							var macUp=String(mac).toUpperCase()
+							if (typeof hosts[macUp] !== 'undefined') {
+								if ((String(hosts[macUp]['ipaddrs'][0]).length>0) && (typeof hosts[macUp]['ipaddrs'][0] !== 'undefined')) {
+									foundname=hosts[macUp]['ipaddrs'][0];
+								}
+								if ((String(hosts[macUp]['name']).length>0) && (typeof hosts[macUp]['name'] !== 'undefined')) {
+									foundname=hosts[macUp]['name'];
+								}
+							}
+							hostl=hostl+ foundname+'&emsp;';
```
[the remark this answers is not in the document]
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:176]: agree

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+							Remoteinfo[wlan]['rrm_nr'][2]
+
+						]);
+				}	
+		cbi_update_table(connectioninfo_table, connectioninfo_table_entries, E('em', _('No data')));
+		body.appendChild(connectioninfo_table);
+/////////////////////////////////////////
+
+
+		var compactconnectioninfo_table = E('table', { 'class': 'table cbi-section-table' }, [
```
[the remark this answers is not in the document]
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:148]: That changes the layout completely. That is not an improvement...
image
image

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/settings.js (hunk 9 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/settings.js:34]: done

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (jow-) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:148]: What theme is this?

[the remark this answers is not in the document]
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:148]: just default

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 2 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:128]: ditto

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 3 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:180]: ditto

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 4 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:208]: ditto

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+		var body = E([
+			E('h2', _('Usteer status'))
+		]);
+
+
+/////////////////////////
+		body.appendChild(
+			E('h3', 'Remotehosts')
+		);
+		var remotehost_table = E('table', { 'class': 'table cbi-section-table' }, [
```
[the remark this answers is not in the document]
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:70]: see above

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+				}
+
+		cbi_update_table(remotehost_table, remotehosttableentries, E('em', _('No data')));
+		body.appendChild(remotehost_table);
+		
+/////////////////////////
+		body.appendChild(
+			E('h3', 'Client list')
+		);
+		var connectioninfo_table = E('table', { 'class': 'table cbi-section-table' }, [
```
[the remark this answers is not in the document]
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:91]: see above

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+							Remoteinfo[wlan]['ssid'],
+							Remoteinfo[wlan]['freq'],
+							Remoteinfo[wlan]['n_assoc'],
+							Remoteinfo[wlan]['noise'],
+							Remoteinfo[wlan]['load'],
+							Remoteinfo[wlan]['max_assoc'],
+							Remoteinfo[wlan]['roam_events']['source'],
+							Remoteinfo[wlan]['roam_events']['target'],
+							Remoteinfo[wlan]['rrm_nr'][0],
+							Remoteinfo[wlan]['rrm_nr'][2]
```
Comment (McGiverGim) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:139]: This is producing an error in my system. It seems `Remoteinfo[wlan]['rrm_nr']` does not exist. I don't know why. It exists in the `Localinfo` but not in the `Remoteinfo`.

image

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (McGiverGim) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:148]: As curiosity I've tested this. With bootstrap theme (default) it seems the design is broken. If I use argon, for example, the design is correct.

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 14 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:139]: ok weird, on my setup it gets filled in, is the "network" correctly filled in on all APs? Regardless let me add some error checking on that, if its "undefined" then i will skip printing

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/status.js:148]: Maybe make an issue for this, such that somebody can look at it?

**Comment by Ramon00** [comment]
im having trouble squashing the commits, if i do that then they dubplicate...

**Comment by Ramon00** [comment]
ah nevermind, i got it

**Comment by rany2** [comment]
I'm having an issue with some items in hearing map showing `IP: ` without any value. I checked the JSON and it seems in some cases ipaddr is an empty array which is object not "undefined". Code seems to check if `hosts..["ipaddr"]` is `undefined` when I think it should be `typeof ...["ipaddr"] !== undefined && ...length > 0`.

image

**Comment by rany2** [comment]
Notes about settings page:

- Missing probe steering, defaults to 0.
- Load balancing threshold defaults to 0, not 5.

See the OpenWRT wiki about this, I recently updated the information on it.

**Comment by Ramon00** [comment]
> 'm having an issue with some items in hearing map showing `IP: ` without any value. I checked the JSON and it seems in some cases ipaddr is an empty array which is object not "undefined". Code seems to check if `hosts..["ipaddr"]` is not `undefined` when I think it should be `typeof ...["ipaddr"] !== undefined && ...length > 0`.
Ok I added a check. Try it again please :)

**Comment by Ramon00** [comment]
> Notes about settings page:
> 
> * Missing probe steering, defaults to 0.
> * Load balancing threshold defaults to 0, not 5.
> 
> See the OpenWRT wiki about this, I recently updated the information on it.

Ok updated. Thx

**Comment by rany2** [comment]
Getting `hosts[macUp].name is undefined` now.

**Comment by Ramon00** [comment]
> @Ramon00
> 
> ```
> 				if ((String(hosts[macUp]['ipaddrs'][0]).length>0) && (typeof hosts[macUp]['ipaddrs'] !== 'undefined'))
> 					maciphost+='\u2003IP: '+hosts[macUp]['ipaddrs'];
> 				if ((String(hosts[macUp]['name'][0]).length>0) && (typeof hosts[macUp]['name'] !== 'undefined')) 
> 					maciphost+='\u2003Host: %h'.format(hosts[macUp]['name']);
> 				}
> ```
> 
> should be
> 
> ```
> 				if ((typeof hosts[macUp]['ipaddrs'] !== 'undefined') && (String(hosts[macUp]['ipaddrs'][0]).length>0))
> 					maciphost+='\u2003IP: '+hosts[macUp]['ipaddrs'];
> 				if ((typeof hosts[macUp]['name'] !== 'undefined') && (String(hosts[macUp]['name'][0]).length>0)) 
> 					maciphost+='\u2003Host: %h'.format(hosts[macUp]['name']);
> 				}
> ```

The [0] with the "name" was a copy/paste mistake, there should be only 1 name in the list, so the index zero is not needed. For the ipadders I think it makes sense to  use 
if ((String(hosts[macUp]['ipaddrs'][0]).length>0) && (typeof hosts[macUp]['ipaddrs'][0] !== 'undefined'))
not sure if its better to rempve the p0[ for the "undefined" check?

**Comment by Ramon00** [comment]
One question to all, does it make sense to break down the "RRM NR hex string"? There should be info in there to indicate capabilities of APs. Not sure if it is interesting information or not.

**Comment by rany2** [comment]
> The [0] with the "name" was a copy/paste mistake, there should be only 1 name in the list, so the index zero is not needed. For the ipadders I think it makes sense to use
> if ((String(hosts[macUp]['ipaddrs'][0]).length>0) && (typeof hosts[macUp]['ipaddrs'][0] !== 'undefined'))
> not sure if its better to rempve the [0] for the "undefined" check?

I think you could just check if the # of array elements instead. I think for most people, only ip6 might have multiple addresses.

> One question to all, does it make sense to break down the "RRM NR hex string"? There should be info in there to indicate capabilities of APs. Not sure if it is interesting information or not.

I would remove it from the status page or hide it behind some debug checkbox. I feel like if someone needs this info, they'd use the CLI but just my opinion.

**Comment by rany2** [comment]
Also I think `rrm nr mac` is redundant as it always be the BSSID.

**Comment by rany2** [comment]
This is a nitpick, but could we instead of what we currently have for client headers:

 `Mac: 22:59:9c:82:3c:6e IP: 192.168.1.2 Host: laptop.home.arpa`

Have this: `laptop.home.arpa (22:59:9c:82:3c:6e, 192.168.1.2, fd00:1234:1234::1)` (similar to the Wireless tab's "Associated Stations" table.

**Comment by Ramon00** [comment]
> I would remove it from the status page or hide it behind some debug checkbox. I feel like if someone needs this info, they'd use the CLI but just my opinion.



> Also I think `rrm nr mac` is redundant as it will always be the BSSID.

Yeah I was thinking it is a bit useless both. Hence the reason to ask. Let me see if I can maybe replace it with some of the info contained in it, like HT,  VHT 
image

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/settings.js [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+	s = m.section(form.TypedSection, 'usteer', _('Options'), _('First four options are mandatory. Also be sure to enable rrm reports, 80211kvr, etc see: https://openwrt.org/docs/guide-user/network/wifi/usteer'));
+	s.anonymous = true;
+
+	s.option(form.Value, 'network', _('Network'), _('The network interface for inter-AP communication)'));
+
+	o = s.option(form.Flag, 'syslog', _('Syslog'), _('Log messages to syslog (0/1)'));
+	o.default = '1';
+	o.rmempty = false;
+
+	o = s.option(form.Flag, 'ipv6', _('Ipv6 mode'), _('Use IPv6 for remote exchange'));
```
Comment (rany2) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/settings.js:20]: Rename Ipv6 => IPv6

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/settings.js [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+'require view';
+'require form';
+
+return view.extend({
+	render: function() {
+	var m, s, o;
+
+	m = new form.Map('usteer', _('Usteer'), _('Usteer configuration.'));
+
+	s = m.section(form.TypedSection, 'usteer', _('Options'), _('First four options are mandatory. Also be sure to enable rrm reports, 80211kvr, etc see: https://openwrt.org/docs/guide-user/network/wifi/usteer'));
```
Comment (rany2) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/settings.js:11]: 802.11r is not required for usteer.

**Comment by rany2** [comment]
> > I would remove it from the status page or hide it behind some debug checkbox. I feel like if someone needs this info, they'd use the CLI but just my opinion.
> 
> > Also I think `rrm nr mac` is redundant as it will always be the BSSID.
> 
> Yeah I was thinking it is a bit useless both. Hence the reason to ask. Let me see if I can maybe replace it with some of the info contained in it, like HT, VHT image

I think you can get away with just exposing `multi-band-operation` from usteer ubus, for 99% of cases that should be enough.

**Comment by Ramon00** [comment]
> 
> I think you can get away with just exposing `multi-band-operation` from ubus usteer connected_clients, for 99% of cases that should be enough.

That info is only available on the AP with the clients, not from e.g. my router.

**Comment by rany2** [comment]
@Ramon00 I think you could get rid of it because the frequency field is already exposed, i think the amount of APs with 5ghz and no VHT is close to zero anyway. Same with HT.

**Comment by Ramon00** [comment]
> h 5ghz and no VHT is close to zero anyway. Same with HT.

Yeah I agree, i will remove. If anybody has some suggestion of something useful to extract from it, let me know.

**Comment by rany2** [comment]
BTW great job just some last comments:

- Something like this would be neat: https://github.com/openwrt/luci/pull/6671#issuecomment-1813214434
- For the connected field for the client table, it would be nice if you also stated "False" for when it's not connected. On mobile, it is hard to follow because there is no text on that field when it's not connected. Just looks odd.
- Splitting the status page into separate pages would be good. Right now it's is a bit overwhelming IMO.

**Comment by Ramon00** [comment]
> * Something like this would be neat: [luci-app-usteer: First version of luci-app-usteer #6671 (comment)](https://github.com/openwrt/luci/pull/6671#issuecomment-1813214434)

well that is possible i guess, but the wireless status page has  "mac, host (IPv4, IPv6), if no host then : "mac, IPv4 IPv6."  if no IP address then  "mac, ?" depending on what info is available. I do think that actually the formatting on the wireless status page is a bit inconsistent as it is. So I am not sure.

> * For the connected field for the client table, it would be nice if you also stated "False" for when it's not connected. On mobile, it is hard to follow because there is no text on that field when it's not connected. Just looks odd.

I had a quick look and made it "yes" "no", as that is more to the point and easier to spot the difference if you scroll through the list quickly. Have a look.

> * Splitting the status page into separate pages would be good. Right now it's is a bit overwhelming IMO.

That would be possible of course. I guess the hearing map would then be the candidate to be moved to another page. But what do other people think? Better to have everything in 1 go, or split? I am actually fine either way

**Comment by rany2** [comment]
> That would be possible of course. I guess the hearing map would then be the candidate to be moved to another page. But what do other people think? Better to have everything in 1 go, or split? I am actually fine either way

I believe DAWN luci app already splits those into separate pages, so you might want to look at that for inspiration.

**Comment by rany2** [comment]
Sorry for another comment, but I was configuring a new node using your new luci app and I noticed that:

- band_steering_interval: Attempting to steer clients to a higher frequency-band every n ms. A value of 0 disabled band-steering. (Default: 120000)
- band_steering_min_snr: Minimal SNR or absolute signal a device has to maintain over band_steering_interval to be steered to a higher frequency band. (Default: -60)

were missing from the settings page.

**Comment by Ramon00** [comment]
> Sorry for another comment,

Ah no problem :) Those must have been added after I made the page initially. Check it now!

**Comment by Ramon00** [comment]
@hnyman  Can you have a look at this PR, and approve it if ok?

**Comment by danpawlik** [comment]
LGTM
```
Tested-by: Daniel Pawlik <dduuch@gmail.com>
```

**Comment by hnyman** [comment]
Some comments. I do not normally use usteer, so a bit hard to test and evaluate ;-)
But I installed usteer to my routers, and this PR seems to work. At least it shows the status, hearing maps etc. 

However, a few things regarding the PR:
* Fails formalities test in CI: The commit title should start with "luci-app-usteer:" (like the PR does).
* There are lots of trailing spaces/tabs after content. At least 40 occurences, I think. They should be removed, just for tidying up.  (git commit should actually warn you about that)

For longer development: 
I do not like that this creates a new top-menu item. It should be part of Services or Network menu.

In general, it looks useful for usteer users.

**Comment by Ramon00** [comment]
@hnyman:
> * Fails formalities test in CI: The commit title should start with "luci-app-usteer:" (like the PR does).
That should be easy to fix

> * There are lots of trailing spaces/tabs after content. At least 40 occurences, I think. They should be removed, just for tidying up.  (git commit should actually warn you about that)
I can do that np.

> For longer development:
> I do not like that this creates a new top-menu item. It should be part of Services or Network menu.
Shall i make it into 1 page with 3 tabs instead?

**Comment by hnyman** [comment]
> Shall i make it into 1 page with 3 tabs instead?

That would be better

**Comment by Ramon00** [comment]
oh hmmm i seem to have broken the settings part. Can anybody spot the mistake?

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+		
+		s = m.section(form.NamedSection, 'hearingmap', 'Hearing map');
+		s.anonymous = true;
+
+		o = s.option(HearingMap);
+
+		s = m.section(form.TypedSection, 'settings','Settings',  _('First four options are mandatory. Also be sure to enable rrm reports, 80211kv, etc see: https://openwrt.org/docs/guide-user/network/wifi/usteer'));
+		s.anonymous = true;
+
+		s.option(form.Value, 'network', _('Network'), _('The network interface for inter-AP communication)'));
```
Comment (hnyman) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:262]: One ) too much?
 But it is inside a string, so it shouldn't hurt, right?

Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:262]: Let me try, but I dont think an extra ) inside a string can be the issue. I think it is related that i switched to` m = new form.JSONMap(mapdata, 'Usteer'); ` instead of `form.Map`, i think that changed the way the uci settings are mapped. If i dont then the different tabs do not work correctly. But I am not sure how to fix it.

Comment (hnyman) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:262]: luci-app-nlbwmon might offer you an example of pretty similar structure. Three pages shown as tabs, one item in menu.

Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js]: I think its fixed. I moved back to a form.Map and tabbed the section instead of the map.  Please have another look

**Comment by Ramon00** [comment]
@hnyman yes, that would be the next thing to fix. I probably will also add refresh of the status when I get a chance. Thx for the help
